validator/pandora/service.go

```diff
@@ -0,0 +1,214 @@
+package pandora
+
+import (
+	"context"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/prysmaticlabs/prysm/shared/logutil"
+	"time"
+
+	eth1Types "github.com/ethereum/go-ethereum/core/types"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+
+var (
+	logThreshold = 8
+	dialInterval = 5 * time.Second
+
+	ConnectionError = errors.New("Client not connected")
+	errNotSynced = errors.New("Pandora node is still syncing")
+	errNoEndpoint = errors.New("No endpoint defined")
+)
+
+type ExtraData struct {
+	Slot 					uint64
+	Epoch   				uint64
+	ProposerIndex 			uint64
+	CoinbaseAddress         common.Address
+}
+
+// Client defines a subset of methods conformed to by Pandora RPC clients for
+// producing catalyst block and insert pandora block.
+type PandoraService interface {
+	// GetWork gets the new block header and hash of pandora client
+	GetWork(ctx context.Context) (*eth1Types.Header, *ExtraData, error)
+	// SubmitWork submits the header hash and signature of pandora block header
+	SubmitWork(ctx context.Context, blockNonce uint64, headerHash common.Hash, sig [32]byte) (bool, error)
+}
+
+type RPCClient interface {
+	Call(result interface{}, method string, args ...interface{}) error
+}
+
+type DialRPCFn func(endpoint string) (*PandoraClient, error)
+
+type Service struct {
+	connected 			  bool
+	isRunning             bool
+	ctx                   context.Context
+	cancel                context.CancelFunc
+	endpoint              string
+	pandoraClient    	  *PandoraClient
+	runError              error
+	dialPandoraFn 		  DialRPCFn
+}
+
+func NewService(ctx context.Context, endpoint string, dialPandoraFn DialRPCFn) (*Service, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	_ = cancel // govet fix for lost cancel. Cancel is handled in service.Stop()
+
+	pandoraClient, err := dialPandoraFn(endpoint)
+	if err != nil {
+		log.WithError(err).Error("Pandora service initialization failed!")
+		return nil, errors.Wrap(err, "Pandora service initialization failed!")
+	}
+
+	return &Service{
+		ctx:              	ctx,
+		cancel:           	cancel,
+		endpoint:      	  	endpoint,
+		dialPandoraFn: 		dialPandoraFn,
+		pandoraClient:      pandoraClient,
+	}, nil
+}
+
+func (s *Service) Start() {
+	go func() {
```

Comment:
Im not quite sure thats the way, `go func()` will die right after `Start()` will return (line 86). Go routines live as long as the invoker. The invoker should start like `go service.Start()` and within body of start (line 78) you should remove it then.

Reply:
https://youtu.be/ARHXmR0_MGY?t=278

Reply:
This works fine, I have checked it. It will live as long as `main()` or `package_name()` function. I have runned your code and experimented with several time.sleeps. Its fine, ignore previous suggestion

Reply:
Thanks.